exec/vector/src/main/java/org/apache/drill/exec/record/metadata/DictBuilder.java

```diff
@@ -0,0 +1,198 @@
+    return this;
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, int width) {
+    tupleBuilder.add(DictVector.FIELD_VALUE_NAME, type, width);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_KEY_NAME, type, TypeProtos.DataMode.REQUIRED, precision, scale);
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_VALUE_NAME, type, TypeProtos.DataMode.REQUIRED, precision, scale);
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type) {
+    tupleBuilder.addNullable(name,  type);
+    return this;
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type, int width) {
+    tupleBuilder.addNullable(name, type, width);
+    return this;
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(name, type, TypeProtos.DataMode.OPTIONAL, precision, scale);
+  }
+
+  public DictBuilder addArrayValue(TypeProtos.MinorType type) {
+    tupleBuilder.addArray(DictVector.FIELD_VALUE_NAME, type);
+    return this;
+  }
+
+  public DictBuilder addArrayValue(TypeProtos.MinorType type, int dims) {
+    tupleBuilder.addArray(DictVector.FIELD_VALUE_NAME, type, dims);
+    return this;
+  }
+
+  public DictBuilder addArrayValue(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_VALUE_NAME, type, TypeProtos.DataMode.REPEATED, precision, scale);
+  }
+
+  public DictBuilder addDecimal(String name, TypeProtos.MinorType type,
+                               TypeProtos.DataMode mode, int precision, int scale) {
+    tupleBuilder.addDecimal(name, type, mode, precision, scale);
+    return this;
+  }
+
+  /**
+   * Add a map column as dict's value. The returned schema builder is for the nested
+   * map. Building that map, using {@link DictBuilder#resumeSchema()},
+   * will return the original schema builder.
+   *
+   * @return a builder for the map
+   */
+  public MapBuilder addMapValue() {
+    return tupleBuilder.addMap(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public MapBuilder addMapArrayValue() {
+    return tupleBuilder.addMapArray(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public DictBuilder addDictValue() {
+    return tupleBuilder.addDict(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public UnionBuilder addUnionValue() {
+    return tupleBuilder.addUnion(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public UnionBuilder addListValue() {
+    return tupleBuilder.addList(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public RepeatedListBuilder addRepeatedListValue() {
+    return tupleBuilder.addRepeatedList(this, DictVector.FIELD_VALUE_NAME);
+  }
+
+  public DictColumnMetadata buildColumn() {
+    return new DictColumnMetadata(memberName, mode, tupleBuilder.schema());
```

Comment:
Sorry, I'm confused. A `DICT` is essentially a tuple? I thought it was an array of (key, value) pairs? Essentially (key: ARRAY<KEY_TYPE>, value: ARRAY<VALUE_TYPE>)?

Let's clarify this before I comment more as I'll need to understand that to make useful comments elsewhere.

It may be that this class was patterned after the Tuple classes. But, if so, then those classes may have been poor examples because a MAP solves a different problem than a DICT (I think...)

Reply:
Indeed, the column metadata was patterned after `TUPLE` classes. This may seem weird, but I see it perfectly fit: `dict` is a "tuple" with two fields (`key`, which has type restrictions, and `value`). This "tuple" is then handled by a special writer, `DictWriter`, and a special reader `DictReader`.